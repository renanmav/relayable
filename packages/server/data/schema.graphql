input AcceptAnswerInput {
  id: ID!
  clientMutationId: String
}

type AcceptAnswerPayload {
  error: String
  answer: Answer
  clientMutationId: String
}

"""Answer data"""
type Answer implements Node & Post {
  """The ID of an object"""
  id: ID!
  _id: String
  question: Question
  content: String
  upvotes: Int
  downvotes: Int
  is_accepted: Boolean
  author: User
  createdAt: String
  updatedAt: String
}

input CreateAnswerInput {
  """ID of the question"""
  question: ID!
  content: String!
  clientMutationId: String
}

type CreateAnswerPayload {
  error: String
  answer: Answer
  clientMutationId: String
}

input CreateQuestionInput {
  title: String!
  content: String!
  tags: [String]
  clientMutationId: String
}

type CreateQuestionPayload {
  error: String
  question: Question
  clientMutationId: String
}

input DeleteQuestionInput {
  id: ID!
  clientMutationId: String
}

type DeleteQuestionPayload {
  error: String
  clientMutationId: String
}

input EditAnswerInput {
  id: ID!
  content: String!
  clientMutationId: String
}

type EditAnswerPayload {
  error: String
  answer: Answer
  clientMutationId: String
}

input EditQuestionInput {
  id: ID!
  title: String
  content: String
  tags: [String]
  clientMutationId: String
}

type EditQuestionPayload {
  error: String
  question: Question
  clientMutationId: String
}

input LoginWithGithubInput {
  code: String!
  clientMutationId: String
}

type LoginWithGithubPayload {
  token: String
  user: User
  clientMutationId: String
}

type Mutation {
  LoginWithGithub(input: LoginWithGithubInput!): LoginWithGithubPayload

  """Use this mutation to create a question"""
  CreateQuestion(input: CreateQuestionInput!): CreateQuestionPayload

  """Mutation to edit the question if logged in user is the author"""
  EditQuestion(input: EditQuestionInput!): EditQuestionPayload

  """Use this mutation to delete a question if you owns it"""
  DeleteQuestion(input: DeleteQuestionInput!): DeleteQuestionPayload

  """
  Mutation to view the question
  Each user represents one view
  """
  ViewQuestion(input: ViewQuestionInput!): ViewQuestionPayload

  """
  Mutation to up/down vote a question
  
  Each user vote only once (and can change his vote)
  """
  VoteQuestion(input: VoteQuestionInput!): VoteQuestionPayload

  """Use this mutation to answer a question"""
  CreateAnswer(input: CreateAnswerInput!): CreateAnswerPayload

  """Mutation to edit the answer if logged in user is the author"""
  EditAnswer(input: EditAnswerInput!): EditAnswerPayload

  """
  Mutation to up/down vote a answer
  
  Each user vote only once (and can change his vote)
  """
  VoteAnswer(input: VoteAnswerInput!): VoteAnswerPayload

  """
  Mutation to accept the answer
  Only the owner of the question can accept the answer
  Only one answer can be accepted
  """
  AcceptAnswer(input: AcceptAnswerInput!): AcceptAnswerPayload
}

type NewQuestionPayload {
  edge(token: String!): QuestionEdge
}

type NewViewPayload {
  edge(token: String!): QuestionEdge
}

type NewVotePayload {
  edge(token: String!): QuestionEdge
}

"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: String

  """When paginating forwards, the cursor to continue."""
  endCursor: String
}

"""Post interface to enforce some fields"""
interface Post {
  content: String
  upvotes: Int
  downvotes: Int
  author: User
  createdAt: String
  updatedAt: String
}

"""The root of all queries"""
type Query {
  """Use this query to fetch GitHub login URL"""
  githubLoginUrl: String

  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID!
  ): Node

  """Get your own data based on provided Authorization token"""
  me: User

  """Fetch a user by its login"""
  user(login: String!): User

  """
  Returns a connection with connectionArgs and/or the Author global ID
  
  You can search questions too with the search arg
  
  Search its done by title, content and tags fields
  """
  questions(after: String, first: Int, before: String, last: Int, authorId: ID, search: String): QuestionConnection

  """
  Returns the average difference between a question creation and its first answer
  
  Value returned is in minutes
  """
  questionAvgResponse: String
}

"""Question data"""
type Question implements Node & Post {
  """The ID of an object"""
  id: ID!
  _id: String
  title: String
  content: String
  upvotes: Int
  downvotes: Int
  views: Int
  anonymous_views: Int
  total_views: Int
  tags: [String]
  author: User
  answers: [Answer]
  createdAt: String
  updatedAt: String
}

"""A connection to a list of items."""
type QuestionConnection {
  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [QuestionEdge]
}

"""An edge in a connection."""
type QuestionEdge {
  """The item at the end of the edge"""
  node: Question

  """A cursor for use in pagination"""
  cursor: String!
}

type Subscription {
  NewQuestion: NewQuestionPayload
  NewView: NewViewPayload
  NewVote: NewVotePayload
}

"""User data"""
type User implements Node {
  """The ID of an object"""
  id: ID!
  _id: String
  github_id: String
  name: String
  login: String
  avatar_url: String
}

input ViewQuestionInput {
  id: ID!
  clientMutationId: String
}

type ViewQuestionPayload {
  error: String
  question: Question
  clientMutationId: String
}

input VoteAnswerInput {
  id: ID!
  up: Boolean
  down: Boolean
  clientMutationId: String
}

type VoteAnswerPayload {
  error: String
  answer: Answer
  clientMutationId: String
}

input VoteQuestionInput {
  id: ID!
  up: Boolean
  down: Boolean
  clientMutationId: String
}

type VoteQuestionPayload {
  error: String
  question: Question
  clientMutationId: String
}
