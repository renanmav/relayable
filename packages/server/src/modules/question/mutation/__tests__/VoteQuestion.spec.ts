import { graphql } from 'graphql'
import { toGlobalId } from 'graphql-relay'
import { PubSub } from 'graphql-subscriptions'

import { schema } from '../../../../schema'
import {
  getContext,
  createRows,
  connectMongoose,
  clearDbAndRestartCounters,
  disconnectMongoose,
} from '../../../../../test/helper'

beforeAll(connectMongoose)

beforeEach(clearDbAndRestartCounters)

afterAll(disconnectMongoose)

const query = `
  mutation voteQuestion(
    $id: ID!
    $up: Boolean
    $down: Boolean
  ) {
    VoteQuestion(input: {
      id: $id
      up: $up
      down: $down
    }) {
      error
      question {
        upvotes
        downvotes
      }
    }
  }
`
const rootValue = {}

it('should not vote if logged off', async () => {
  const user = await createRows.createUser()
  const question = await createRows.createQuestion({ author: user._id })
  const context = getContext()
  const variables = {
    id: toGlobalId('Question', question._id),
    up: true,
  }

  const result = await graphql(schema, query, rootValue, context, variables)

  expect(result).toMatchSnapshot()
})

it('should remove his vote', async () => {
  const user = await createRows.createUser()
  const question = await createRows.createQuestion({ author: user._id })
  const pubSub = new PubSub()
  const context = getContext({ user, pubSub })
  const variables = {
    id: toGlobalId('Question', question._id),
  }

  await graphql(schema, query, rootValue, context, { ...variables, up: true })

  const result = await graphql(schema, query, rootValue, context, variables)
  expect(result).toMatchSnapshot()
})

describe('when votting up', () => {
  it('should vote only once', async () => {
    const user = await createRows.createUser()
    const question = await createRows.createQuestion({ author: user._id })
    const pubSub = new PubSub()
    const context = getContext({ user, pubSub })
    const variables = {
      id: toGlobalId('Question', question._id),
      up: true,
    }

    const result = await graphql(schema, query, rootValue, context, variables)
    expect(result).toMatchSnapshot()

    const secondResult = await graphql(schema, query, rootValue, context, variables)
    expect(secondResult).toMatchSnapshot()
  })

  it('should remove down vote if so', async () => {
    const user = await createRows.createUser()
    const question = await createRows.createQuestion({ author: user._id })
    const pubSub = new PubSub()
    const context = getContext({ user, pubSub })
    const variables = {
      id: toGlobalId('Question', question._id),
    }

    const result = await graphql(schema, query, rootValue, context, {
      ...variables,
      down: true,
    })
    expect(result).toMatchSnapshot()

    const secondResult = await graphql(schema, query, rootValue, context, {
      ...variables,
      up: true,
    })
    expect(secondResult).toMatchSnapshot()
  })
})

describe('when votting down', () => {
  it('should vote only once', async () => {
    const user = await createRows.createUser()
    const question = await createRows.createQuestion({ author: user._id })
    const pubSub = new PubSub()
    const context = getContext({ user, pubSub })
    const variables = {
      id: toGlobalId('Question', question._id),
      down: true,
    }

    const result = await graphql(schema, query, rootValue, context, variables)
    expect(result).toMatchSnapshot()

    const secondResult = await graphql(schema, query, rootValue, context, variables)
    expect(secondResult).toMatchSnapshot()
  })

  it('should remove up vote if so', async () => {
    const user = await createRows.createUser()
    const question = await createRows.createQuestion({ author: user._id })
    const pubSub = new PubSub()
    const context = getContext({ user, pubSub })
    const variables = {
      id: toGlobalId('Question', question._id),
    }

    const result = await graphql(schema, query, rootValue, context, {
      ...variables,
      up: true,
    })
    expect(result).toMatchSnapshot()

    const secondResult = await graphql(schema, query, rootValue, context, {
      ...variables,
      down: true,
    })
    expect(secondResult).toMatchSnapshot()
  })
})
