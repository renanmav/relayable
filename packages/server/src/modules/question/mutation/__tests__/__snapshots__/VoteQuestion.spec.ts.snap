// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not vote if logged off 1`] = `
Object {
  "data": Object {
    "VoteQuestion": Object {
      "error": "You must be authenticated",
      "question": null,
    },
  },
}
`;

exports[`should remove his vote 1`] = `
Object {
  "data": Object {
    "VoteQuestion": Object {
      "error": null,
      "question": Object {
        "downvotes": 0,
        "upvotes": 0,
      },
    },
  },
}
`;

exports[`when votting down should remove up vote if so 1`] = `
Object {
  "data": Object {
    "VoteQuestion": Object {
      "error": null,
      "question": Object {
        "downvotes": 0,
        "upvotes": 1,
      },
    },
  },
}
`;

exports[`when votting down should remove up vote if so 2`] = `
Object {
  "data": Object {
    "VoteQuestion": Object {
      "error": null,
      "question": Object {
        "downvotes": 1,
        "upvotes": 0,
      },
    },
  },
}
`;

exports[`when votting down should vote only once 1`] = `
Object {
  "data": Object {
    "VoteQuestion": Object {
      "error": null,
      "question": Object {
        "downvotes": 1,
        "upvotes": 0,
      },
    },
  },
}
`;

exports[`when votting down should vote only once 2`] = `
Object {
  "data": Object {
    "VoteQuestion": Object {
      "error": "You already have voted down on this question",
      "question": null,
    },
  },
}
`;

exports[`when votting up should remove down vote if so 1`] = `
Object {
  "data": Object {
    "VoteQuestion": Object {
      "error": null,
      "question": Object {
        "downvotes": 1,
        "upvotes": 0,
      },
    },
  },
}
`;

exports[`when votting up should remove down vote if so 2`] = `
Object {
  "data": Object {
    "VoteQuestion": Object {
      "error": null,
      "question": Object {
        "downvotes": 0,
        "upvotes": 1,
      },
    },
  },
}
`;

exports[`when votting up should vote only once 1`] = `
Object {
  "data": Object {
    "VoteQuestion": Object {
      "error": null,
      "question": Object {
        "downvotes": 0,
        "upvotes": 1,
      },
    },
  },
}
`;

exports[`when votting up should vote only once 2`] = `
Object {
  "data": Object {
    "VoteQuestion": Object {
      "error": "You already have voted up on this question",
      "question": null,
    },
  },
}
`;
